Valerie Jimenez and Nick Garcia
Lab 6 Radom Forest
 https://www.kaggle.com/datasets/urvishvekariya/credit-risk-assessment/data

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn import metrics
from sklearn.preprocessing import LabelEncoder

In [ ]:
#read dataset
df=pd.read_csv('credit_risk.csv')
df.head(20)

,person_age,person_income,person_home_ownership,person_emp_length,loan_intent,loan_amnt,loan_int_rate,loan_status,loan_percent_income,cb_person_default_on_file,cb_person_cred_hist_length
0,22,59000,RENT,123.0,PERSONAL,35000,16.02,1,0.59,Y,3
1,21,9600,OWN,5.0,EDUCATION,1000,11.14,0,0.10,N,2
2,25,9600,MORTGAGE,1.0,MEDICAL,5500,12.87,1,0.57,N,3
3,23,65500,RENT,4.0,MEDICAL,35000,15.23,1,0.53,N,2
4,24,54400,RENT,8.0,MEDICAL,35000,14.27,1,0.55,Y,4
5,21,9900,OWN,2.0,VENTURE,2500,7.14,1,0.25,N,2
6,26,77100,RENT,8.0,EDUCATION,35000,12.42,1,0.45,N,3
7,24,78956,RENT,5.0,MEDICAL,35000,11.11,1,0.44,N,4
8,24,83000,RENT,8.0,PERSONAL,35000,8.90,1,0.42,N,2
9,21,10000,OWN,6.0,VENTURE,1600,14.74,1,0.16,N,3


In [ ]:
#any null variables?
df.isnull().sum()

,0
person_age,0
person_income,0
person_home_ownership,0
person_emp_length,895
loan_intent,0
loan_amnt,0
loan_int_rate,3116
loan_status,0
loan_percent_income,0
cb_person_default_on_file,0


In [ ]:
#drop null data
df.dropna(inplace=True)
df.isnull().sum()

,0
person_age,0
person_income,0
person_home_ownership,0
person_emp_length,0
loan_intent,0
loan_amnt,0
loan_int_rate,0
loan_status,0
loan_percent_income,0
cb_person_default_on_file,0


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 28638 entries, 0 to 32580
Data columns (total 11 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   person_age                  28638 non-null  int64  
 1   person_income               28638 non-null  int64  
 2   person_home_ownership       28638 non-null  object 
 3   person_emp_length           28638 non-null  float64
 4   loan_intent                 28638 non-null  object 
 5   loan_amnt                   28638 non-null  int64  
 6   loan_int_rate               28638 non-null  float64
 7   loan_status                 28638 non-null  int64  
 8   loan_percent_income         28638 non-null  float64
 9   cb_person_default_on_file   28638 non-null  object 
 10  cb_person_cred_hist_length  28638 non-null  int64  
dtypes: float64(3), int64(5), object(3)
memory usage: 2.6+ MB


In [ ]:
#label encoder for strings
le=LabelEncoder()
df['person_home_ownership']=le.fit_transform(df['person_home_ownership'])
df['loan_intent']=le.fit_transform(df['loan_intent'])
df['cb_person_default_on_file']=le.fit_transform(df['cb_person_default_on_file'])
df.head(20)

,person_age,person_income,person_home_ownership,person_emp_length,loan_intent,loan_amnt,loan_int_rate,loan_status,loan_percent_income,cb_person_default_on_file,cb_person_cred_hist_length
0,22,59000,3,123.0,4,35000,16.02,1,0.59,1,3
1,21,9600,2,5.0,1,1000,11.14,0,0.10,0,2
2,25,9600,0,1.0,3,5500,12.87,1,0.57,0,3
3,23,65500,3,4.0,3,35000,15.23,1,0.53,0,2
4,24,54400,3,8.0,3,35000,14.27,1,0.55,1,4
5,21,9900,2,2.0,5,2500,7.14,1,0.25,0,2
6,26,77100,3,8.0,1,35000,12.42,1,0.45,0,3
7,24,78956,3,5.0,3,35000,11.11,1,0.44,0,4
8,24,83000,3,8.0,4,35000,8.90,1,0.42,0,2
9,21,10000,2,6.0,5,1600,14.74,1,0.16,0,3


In [ ]:
df.columns

Index(['person_age', 'person_income', 'person_home_ownership',
       'person_emp_length', 'loan_intent', 'loan_amnt', 'loan_int_rate',
       'loan_status', 'loan_percent_income', 'cb_person_default_on_file',
       'cb_person_cred_hist_length'],
      dtype='object')

In [ ]:
#split dataset in features and target variable
x_data=df[['person_age', 'person_income', 'person_home_ownership',
       'person_emp_length', 'loan_intent', 'loan_amnt', 'loan_int_rate', 'loan_percent_income', 'cb_person_default_on_file',
       'cb_person_cred_hist_length']]
y_data=df['loan_status']

In [ ]:
# Split dataset into training set and test set
X_train, X_test, y_train, y_test = train_test_split(x_data, y_data, test_size=0.3, random_state=1)


In [ ]:
#train a single Decision tree
dt_model=DecisionTreeClassifier(random_state=42)
dt_model.fit(X_train, y_train)

# Calculate the accuracy and assign it to dt_accuracy
dt_accuracy = dt_model.score(X_test, y_test)

print("Decision Tree Test Accuracy:", f"{dt_accuracy * 100:.2f}%")


Decision Tree Test Accuracy: 86.85%


In [ ]:

#train a Random Forest with random feature selection
rf_model = RandomForestClassifier(n_estimators=5, max_features=2, random_state=42)  # max_features=2 means only 2 features are considered at each split


rf_model.fit(X_train, y_train)


#calculate the accuracy and assign it to rf_accuracy
rf_accuracy = rf_model.score(X_test, y_test)

# Now you can use rf_accuracy in the print statements below
print("RF Model Accuracy:", f"{rf_accuracy * 100:.2f}", "%")



RF Model Accuracy: 89.82 %


In [ ]:
# Make predictions on the test set
y_pred_rf = rf_model.predict(X_test)

# Calculate precision, recall, and F1-score for Random Forest
precision_rf = metrics.precision_score(y_test, y_pred_rf)
recall_rf = metrics.recall_score(y_test, y_pred_rf)


f1_rf = metrics.f1_score(y_test, y_pred_rf)


print("RF Precision:", precision_rf)
print("RF Recall:", recall_rf)
print("RF F1-Score:", f1_rf)

RF Precision: 0.8429172510518934
RF Recall: 0.6486778197517539
RF F1-Score: 0.7331503507166819


In [ ]:
# Prepare a provision for a new data entry and its prediction of an outcome.

def predict_loan():
    while True:
      try:
        person_age=int(input("Enter Age:"))
        person_income=int(input("Enter Income:"))
        person_home_ownership = int(input("Enter Home Ownership(0=Rent, 2=Own, 3=Mortgage, 1=Other): "))
        person_emp_length = float(input("Enter Employment Length"))
        loan_intent = int(input("Enter Loan Intent(1=Educational, 0=Venture, 2=Home, 4=Personal, 3=Medical, 5=Debt):"))
        loan_amnt = int(input("Enter Loan Amount:"))
        loan_int_rate = float(input("Enter Loan Interest Rate float:"))
        loan_percent_income = float(input("Enter Loan % Income float:"))
        cb_person_default_on_file = int(input("Is your account on file? (0=No,1=Yes):"))
        cb_person_cred_hist_length = int(input("Enter Credit History Length:"))

            # Input validation (example: check for non-negative values)
        if any(val < 0 for val in [person_age, person_income, person_home_ownership, person_emp_length, loan_intent, loan_amnt, loan_int_rate, loan_percent_income, cb_person_default_on_file, cb_person_cred_hist_length]):
                print("Error: Please enter non-negative values for all features.")
                continue  #restart the loop

        new_credit_data = pd.DataFrame([[person_age, person_income, person_home_ownership,
       person_emp_length, loan_intent, loan_amnt, loan_int_rate, loan_percent_income, cb_person_default_on_file,
       cb_person_cred_hist_length]], columns=['person_age', 'person_income', 'person_home_ownership', 'person_emp_length', 'loan_intent', 'loan_amnt', 'loan_int_rate', 'loan_percent_income', 'cb_person_default_on_file', 'cb_person_cred_hist_length'])

        prediction =rf_model.predict(new_credit_data)[0]

        if prediction == 1:
                print("Prediction: Approved for Loan.")
        else:
                print("Prediction: Denied Loan.")

        another_loan = input("Enter data for another loan? (yes/no): ")
        if another_loan.lower() != 'yes':
                break #exit loop
      except ValueError:
            print("Invalid input. Please enter valid numerical values.")

predict_loan()



Enter Age:23
Enter Income:120000
Enter Home Ownership(0=Rent, 2=Own, 3=Mortgage, 1=Other): 0
Enter Employment Length2
Enter Loan Intent(1=Educational, 0=Venture, 2=Home, 4=Personal, 3=Medical, 5=Debt):1
Enter Loan Amount:1200
Enter Loan Interest Rate float:12.3
Enter Loan % Income float:.33
Is your account on file? (0=No,1=Yes):0
Enter Credit History Length:2
Prediction: Denied Loan.
Enter data for another loan? (yes/no): no


In [ ]:
# For the feature importance, show a table of features and its corresponding values

feature_importances = pd.DataFrame({'feature': x_data.columns, 'importance': rf_model.feature_importances_})
feature_importances = feature_importances.sort_values('importance', ascending=False)
feature_importances

,feature,importance
6,loan_int_rate,0.214060
7,loan_percent_income,0.203211
1,person_income,0.163402
5,loan_amnt,0.097386
2,person_home_ownership,0.091081
3,person_emp_length,0.064367
4,loan_intent,0.063164
0,person_age,0.046037
9,cb_person_cred_hist_length,0.041078
8,cb_person_default_on_file,0.016214


 A Random Forest classification model was used to analyze the credit risk assessment dataset and predict loan status. The feature variables (X_data) were age, income, home ownership, employment length, loan intent, loan amount, interest rate, percentage of income used for the loan, previous default history, and credit history length. The target variable (y_data) was the loan status, which indicates whether the loan was approved or  denied/considered risky. Creating multiple decision trees and combining their results improved prediction accuracy and reduced overfitting. The Decision Tree Test Accuracy was 86.85% and the Random Forest Model Accuracy was 89.82 %. New data entry for loan status predictions were created and the feature importance value means the higher the value the more impact to the outcome.